### DISCUSSION TITLE
In-editor on-screen keyboard

### DISCUSSION THREAD

**Discussion by roalyr** [godotengine/godot-proposals#7639] [Editor]
# EDIT: Updated link https://github.com/roalyr/CustomKeyboardEngine/

I have seen the plugin for 3.x that implements it, and thought that it may be re-used for the editor module:
https://godotengine.org/asset-library/asset/1328

The idea is that for large-screen device in landscape mode it could be handier to use a keyboard with one thumb either on left or right side, like what I am doing with my custom app:
![IMG_20230910_162551_276.jpg](https://github.com/godotengine/godot-proposals/assets/46825200/20ff6e74-a277-4d98-8507-0460b5e3159b)

When it is stretched on the bottom of the screen it takes away space and isn't handy to use when you hold your tablet with both hands. So in this case you have a phone-sized keyboard suited for the same thumb-typing.

**Comment by roalyr** [comment]
The layout that you see I have been developing and finetuning since 2018, when I started coding on my phone.

**Comment by Zireael07** [comment]
Related to [#76803 ](https://github.com/godotengine/godot/issues/76803)

Can you share the keyboard app?

**Reply by roalyr** [reply]
https://github.com/roalyr/Simple-Keyboard-7row/releases/tag/s7row-keyboard-tablet-1.1.1

https://youtu.be/OCTiacmVu_Y

**Comment by roalyr** [comment]
It is worth mentioning that keyboard text-editing shortcuts (Ctrl+key) do not really work (even with gboard), so having a built-in editor keyboard would be the best UX, in terms of layout as well.

**Comment by roalyr** [comment]
This thing works well too. If you have a second Android device, you can turn it into a BT keyboard + mouse: https://play.google.com/store/apps/details?id=io.appground.blek

**Comment by theludovyc** [comment]
Hello 👋

Maybe you can be interested by this issue https://github.com/godotengine/godot-proposals/issues/10140 😊

**Comment by roalyr** [comment]
Earlier this year I have started writing a new keyboard app that has floating capability like gboard. Base is done, but needs all the keys and features implemented, but I don't really have much time for it now.
https://youtu.be/vBjjByVrKYM?si=9y47n-hPCmjhYvED

**Comment by roalyr** [comment]
I have rolled out a basic release:

https://github.com/roalyr/New7rowKB/releases/tag/1.0
Would appreciate testing and input.

![screen_floating_landscape](https://github.com/user-attachments/assets/8244b045-a097-4ea9-b361-a996c761d972)

**Comment by roalyr** [comment]
This is finally stable (needs testing on different Android versions) and usable (custom json keyboard layouts are working and you can have as many as you want). 
https://github.com/roalyr/CustomKeyboardEngine/releases/tag/1.2.1

**Reply by roalyr** [reply]
![IMG_20241222_131512_428.jpg](https://github.com/user-attachments/assets/4c98ba83-b432-441f-b5fe-4619f7c47c7b)

![IMG_20241222_131512_582.jpg](https://github.com/user-attachments/assets/520bd70d-0c46-4e12-a8e5-2625fd0c0231)
